module/synchronization/core_rapid_test.go

```diff
@@ -77,35 +93,78 @@ func (r *rapidSync) HandleByID(t *rapid.T) {
+	// we collect the received blocks as determined above
+	var receivedBlocks []flow.Header
+	// we also collect the pending blocks
+	var activeBlocks []flow.Header
+
+	// we check the validity of our pushdown automaton for ID requests and populate activeBlocks / receivedBlocks
+	for id, count := range r.idRequests {
+		s, foundID := r.core.blockIDs[id]
+
+		block, foundBlock := findHeader(r.store, func(h flow.Header) bool {
+			return h.ID() == id
+		})
+		require.True(t, foundBlock, "incorrect management of idRequests in the tests: all added IDs are supposed to be from the store")
+
+		if count == 1 {
+			require.True(t, foundID, "ID %v is supposed to be known, but isn't", id)
+
+			assert.True(t, s.WasQueued(), "ID %v was expected to be Queued and is %v", id, s.StatusString())
+			assert.False(t, s.WasReceived(), "ID %v was expected to be Queued and is %v", id, s.StatusString())
+			activeBlocks = append(activeBlocks, *block)
+		} else if count == 0 {
+			if foundID {
+				// if a block is known with 0 pendings, it's because it was received
+				assert.True(t, s.WasReceived(), "ID %v was expected to be Received and is %v", id, s.StatusString())
 			}
+			receivedBlocks = append(receivedBlocks, *block)
 		} else {
-			t.Fatalf("incorrect management of idRequests in the tests")
+			t.Fatalf("incorrect management of idRequests in the tests (0 <= count <= 1)")
 		}
 	}
-	heights, blockIDs := r.core.getRequestableItems()
-	// the queueing logic queues intervals, while our r.heightRequests only queues specific requests
+
+	// we collect still-active heights
 	var activeHeights []uint64
-	for k, v := range r.heightRequests {
-		if v == 1 {
-			activeHeights = append(activeHeights, k)
+
+	for h, count := range r.heightRequests {
+		s, ok := r.core.heights[h]
+		if count == 1 {
+			require.True(t, ok, "Height %x is supposed to be known, but isn't", h)
+			assert.True(t, s.WasQueued(), "Height %x was expected to be Queued and is %v", h, s.StatusString())
+			assert.False(t, s.WasReceived(), "Height %x was expected to be Queued and is %v", h, s.StatusString())
+			activeHeights = append(activeHeights, h)
+
+		} else if count == 0 {
+			// if a height is known with 0 pendings, it's because:
+			// - it was received
+			// - or because a block at this height was (blockAtHeightWasReceived)
+			// - or because a request for a block at that height made us "forget" the prior height reception (clobberedByID)
+			if ok {
+				wasReceived := s.WasReceived()
+				_, blockAtHeightWasReceived := findHeader(receivedBlocks, func(header flow.Header) bool {
+					return header.Height == h
+				})
+				_, clobberedByID := findHeader(activeBlocks, func(header flow.Header) bool {
+					return header.Height == h
+				})
+				heightWasCanceled := wasReceived || blockAtHeightWasReceived || clobberedByID
+
+				assert.True(t, heightWasCanceled, "Height %x was expected to be Received (or filled through a same-height block) and is %v", h, s.StatusString())
```

Comment:
Should we track the time of requests for blocks? Otherwise I think the "clobberedByID" part could give us a false positive. It could return true even if the request for a block at that height happened *before* the height reception.

But actually in this case, we want `heightWasCanceled` to be false.

Reply:
> It could return true even if the request for a block at that height happened before the height reception.
> But actually in this case, we want heightWasCanceled to be false.

So, as this check runs on every single instruction, I'm not sure I see what you mean. If I first have a request for a block at height h, thereby clobbering all height requests at h, and then I receive at height h, then `wasReceived` becomes true.

Perhaps the adjective `canceled` is misleading, and should be replaced by `heightWasReceivedOrCanceled`?